### ISSUE TITLE
Stream state processing errors caused by special characters

### ISSUE THREAD

**Issue by moyiun09** [haproxy/haproxy#2548]
### Detailed Description of the Problem

Unusual frame sequences containing special characters caused haproxy to handle errors, ignore or reset the stream's state transition processing, and forward the request/response.

frames info(special character 0x1a):
```
{
    "<headers-frame-pool-1:1>": {
        "headers_block":{
        ":method": "POST",
        ":scheme": "http",
        ":path": "/flag1",
        ":authority": "localhost:8013",
        "\u001asome-header":"some-value"
        },
        "flags": [
            "ES",
            "EH"
        ],
        "stream_id":"1"
    },
    "<headers-frame-base-1:1>": {
        "headers_block":{
        ":method": "GET",
        ":scheme": "http",
        ":path": "/flag2",
        ":authority": "localhost:8013"
        },
        "flags": [
            "EH"
        ],
        "stream_id":"1"
    },
    "<data-frame-base-2:1>": {
        "data": "5\r\nBBBBB\r\n0\r\n\r\n",
        "flags": ["ES"],
        "stream_id":"1"
    }
}
```

![haproxy_stream_state_trans_2](https://github.com/haproxy/haproxy/assets/62366932/71b24b62-01e5-49d8-8266-8585e92a5e7c)

### Expected Behavior

For abnormal frame sequences that do not contain special characters (0x0a), haproxy handles them as follows.

frames info:
```
{
    "<headers-frame-pool-1:1>": {
        "headers_block":{
        ":method": "POST",
        ":scheme": "http",
        ":path": "/flag1",
        ":authority": "localhost:8013",
        "some-header":"some-value"
        },
        "flags": [
            "ES",
            "EH"
        ],
        "stream_id":"1"
    },
    "<headers-frame-base-1:1>": {
        "headers_block":{
        ":method": "GET",
        ":scheme": "http",
        ":path": "/flag2",
        ":authority": "localhost:8013"
        },
        "flags": [
            "EH"
        ],
        "stream_id":"1"
    },
    "<data-frame-base-2:1>": {
        "data": "5\r\nBBBBB\r\n0\r\n\r\n",
        "flags": ["ES"],
        "stream_id":"1"
    }
}
```

![haproxy_stream_state_trans](https://github.com/haproxy/haproxy/assets/62366932/c16ef808-94e4-4773-a76d-999aa39a2295)

### Steps to Reproduce the Behavior

1. Build a corresponding environment
2. Send a specific frame sequence to reproduce the problem
3. Replace other characters that do not comply with RFC specifications

### Do you have any idea what may have caused this?

haproxy ignores or resets state transition of unfinished stream due to special characters.

### What is your configuration?

```haproxy
defaults
   mode http
   timeout http-keep-alive 10s
   timeout connect 5s
   timeout server 60s
   timeout client 30s
   timeout http-request 30s

frontend http_in
   bind *:8012
    # 配上证书
   #  bind :443 ssl crt /usr/local/haproxy/site.pem alpn h2
   mode http
   timeout client 5s
   timeout http-request 10s
   default_backend upsteam_servers

backend upsteam_servers
   mode http
   http-reuse always
   #fill you host server ip
   server server1 server-ip:80 maxconn 1000
```

### Output of `haproxy -vv`

```plain
moyiun09@moyiun09-VirtualBox:/opt/server/haproxy-2.5.14$ uname -a
Linux moyiun09-VirtualBox 5.15.0-105-generic #115~20.04.1-Ubuntu SMP Mon Apr 15 17:33:04 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
moyiun09@moyiun09-VirtualBox:/opt/server/haproxy-2.5.14$ sudo ./sbin/haproxy  -vv
HAProxy version 2.5.14-52b556a 2023/05/02 - https://haproxy.org/
Status: End of life - please upgrade to branch 2.6.
Known bugs: http://www.haproxy.org/bugs/bugs-2.5.14.html
Running on: Linux 5.15.0-105-generic #115~20.04.1-Ubuntu SMP Mon Apr 15 17:33:04 UTC 2024 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE=1 USE_OPENSSL=1 USE_ZLIB=1
  DEBUG   = 

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL +EPOLL -EVPORTS +GETADDRINFO -KQUEUE +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY -LUA -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OT +PCRE -PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL -PROMEX -QUIC +RT -SLZ -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL +ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 1.0.1u  22 Sep 2016
Running on OpenSSL version : OpenSSL 1.0.1u  22 Sep 2016
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Support for malloc_trim() is enabled.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.39 2016-06-14
Running on PCRE version : 8.39 2016-06-14
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.4.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : none

Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

**Comment by moyiun09** [comment]
env info:
client:     192.168.153.1
server:    192.168.153.128
haproxy: 192.168.153.129

client(h2) --- haproxy(h2/h1) --- server(h1)

**Comment by capflam** [comment]
The 2.5 is end-of-life. You must upgrade, for instance to 2.6.17.

**Comment by capflam** [comment]
Note: the fix for the 2.6 is 54f53ef7c
